go/arrow/array/string.go

```diff
@@ -113,6 +113,12 @@ func (a *String) setData(data *Data) {
 	if offsets := data.buffers[1]; offsets != nil {
 		a.offsets = arrow.Int32Traits.CastFromBytes(offsets.Bytes())
 	}
+
+	expNumOffsets := a.array.data.offset + a.array.data.length + 1
+	if a.array.data.length > 0 &&
+		(len(a.offsets) < expNumOffsets || int(a.offsets[expNumOffsets-2]) > len(a.values)) {
+		panic("arrow/array: string offsets out of bounds of data buffer")
+	}
```

Comment:
is the `a.array.data.length > 0` actually needed? Also, shouldn't this be `len(a.offsets) >= expNumOffsets` ?

`a.offsets[expNumOffsets-2]` would be incorrect, why the -2? `a.offsets[expNumOffset]` should be the final byte.

Reply:
The length check is needed because if length is 0 it appears that lots of things allow `len(a.offsets)` to also be `0`. The offsets seem to be ignored in that case.

I think `len(a.offsets) < expNumOffsets` is correct. If there aren't at least `expNumOffsets` we should panic. If there are extra offsets we can't panic because that can happen in slice scenarios like `array.NewSliceData` (unit tests seem to confirm this).

`a.offsets[expNumOffsets-2]` is where the last value in the array starts and it must be less than `len(a.values)`. I'll change the `>` to a `>=`. The final byte, indicated by `a.offsets[expNumOffsets-1]` seems to be allowed to overflow and be gracefully handled, based on [this test](https://github.com/apache/arrow/blob/master/go/arrow/array/concat_test.go#L293), which is why I am not checking it.

Reply:
Ah I flipped the condition in my head, the `>=` is what indicates it's valid and as such `<` is telling that it's invalid. That's my bad.

Looking at the test you're referencing however, it's considered an *error* if it overflows like that. The fact that arrays can be constructed easily by putting the buffers together can result in such an overflow occurring (because right now we don't validate it) but notice that it's expected that the call to `Concatenate` would return an error in such a case because that is not a valid array. If we're going to panic if someone tries constructing such an invalid array then we can probably remove that test since the situation could never occur (which also means this should probably have a corresponding change in `binary.go` for `Binary` arrays.

If we allow the last offset to go past the end of the data slice, it doesn't solve the original panic you were trying to prevent in the first place as calling `ValueBytes` would still result in an out of bounds panic in that situation.

Reply:
Are you okay with the `Concatenate` error becoming a panic?

I can remove the logic that allows the last offset to overflow and apply this fix to binary arrays as well.

Also, I said I would change `int(a.offsets[expNumOffsets-2]) > len(a.values)` to `int(a.offsets[expNumOffsets-2]) >= len(a.values)` but that doesn't work if the last value in the string array is empty string. In that case the last offset will be equal to `len(a.values)`, so I'm going to leave it as I had it.

> If we allow the last offset to go past the end of the data slice, it doesn't solve the original panic you were trying to prevent in the first place as calling ValueBytes would still result in an out of bounds panic in that situation.

It actually does still solve my original panic. My original issue would be caught because the second-to-last offset was also past the end.

Reply:
I'd prefer the Concatenate error to still be an error if possible, but I'm okay with it being a panic. That said, I don't want to rely on the second-to-last offset, we should follow the Arrow spec and either rely on the last offset or not add this check at all. Alternately we could follow the C++ path and add a Validate function to the arrays and then check that in the reader so we can return an error. 

Either way, I don't like relying on the second-to-last offset